### ISSUE TITLE
Spurious AIP-192 (has-comments) violations

### ISSUE THREAD

**Issue by DRayX** [googleapis/api-linter#714]
I'm getting reports of missing comments over nested messages and their fields.  I have an example that looks something like:

```proto
// A representation of a book.
message Book {
  // The resource name of the book.
  string name = 1;

  // A representation of a page in the book.
  message Page {
    // The content of the page.
    string content = 1;
  }

  // The pages of the book.
  repeated Page pages = 2;
}
```

But I get the equivalent of `Missing comment over "Book"` and `Missing comment over "content"`even though both the message and the fields have comments.

**Comment by lukesneeringer** [comment]
Looks like a quirk of nested messages. Will look into it.

**Comment by apasel422** [comment]
I'm not able to reproduce with the example from https://github.com/googleapis/api-linter/issues/714#issue-747923427.

**Comment by apasel422** [comment]
I can reproduce with a different example. Investigating further...

**Comment by apasel422** [comment]
It appears that the key being used to store the source info is different from the one being used to look it up.

**Comment by apasel422** [comment]
I wonder if this is because `map` fields end up being translated into a nested message, and there's something wrong with the offset there...

**Comment by apasel422** [comment]
Yes, that's it. I can reproduce with the following example

```go
// A representation of a book.
message Book {
  // The resource name of the book.
  string name = 1;

  // this map ends up being represented as a nested message, which
  // throws off the source location path in the protoreflect library
  map<int32, string> some_map = 2;

  // A representation of a page in the book.
  message Page {
    // The content of the page.
    string content = 1;
  }

  // The pages of the book.
  repeated Page pages = 3;
}
```

**Comment by apasel422** [comment]
CC @jhump.

**Comment by apasel422** [comment]
I think the issue is that protoreflect's `MessageDescriptor.resolve` is [using](https://github.com/jhump/protoreflect/blob/8269465cdf82d93fb4ec1cd601d5224f9624418a/desc/descriptor.go#L328-L329) the nested type's index in the `DescriptorProto.nested_types` field to generate the path into the source code info, but there is no source code location for the synthetic nested map entry message.

**Comment by apasel422** [comment]
I think the issue is with `protoparse`, as the file descriptors generated by `protoc` seem to work just fine.
